okhttp/src/main/java/io/grpc/transport/okhttp/OutboundFlowController.java

```diff
@@ -94,24 +95,36 @@ synchronized void windowUpdate(int streamId, int delta) {
       writeStreams();
     } else {
       // Update the stream window and write any pending frames for the stream.
-      OutboundFlowState state = stateOrFail(streamId);
-      state.incrementStreamWindow(delta);
+      OutboundFlowState state = state(streamId);
+      if (state != null) {
+        state.incrementStreamWindow(delta);
 
-      WriteStatus writeStatus = new WriteStatus();
-      state.writeBytes(state.writableWindow(), writeStatus);
-      if (writeStatus.hasWritten()) {
-        flush();
+        WriteStatus writeStatus = new WriteStatus();
+        state.writeBytes(state.writableWindow(), writeStatus);
+        if (writeStatus.hasWritten()) {
+          flush();
+        }
+      } else if (!transport.mayHaveCreatedStream(streamId)) {
+        transport.onError(ErrorCode.PROTOCOL_ERROR,
+            "Received window_update for unknown stream: " + streamId);
       }
     }
   }
 
   synchronized void data(boolean outFinished, int streamId, Buffer source, boolean flush) {
     Preconditions.checkNotNull(source, "source");
-    if (streamId <= 0) {
-      throw new IllegalArgumentException("streamId must be > 0");
+    if (streamId <= 0 || !transport.mayHaveCreatedStream(streamId)) {
+      throw new IllegalArgumentException("Invalid streamId: " + streamId);
+    }
+
+    OutboundFlowState state = state(streamId);
+    if (state == null) {
+      // This is possible for a stream that has received end-of-stream from server (but hasn't sent
+      // end-of-stream), and was removed from the transport stream map.
+      // In such case, we just throw away the data.
+      return;
     }
 
-    OutboundFlowState state = stateOrFail(streamId);
```

Comment:
No, this is called by OkHttpClientStream (https://github.com/grpc/grpc-java/blob/master/okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientStream.java#L179) for outbound messages, the stream might be removed as I described in the comment.